### ISSUE TITLE
in_array 3rd param true - TypeDoesNotContainType error but not vice versa

### ISSUE THREAD

**Issue by kkmuffme** [vimeo/psalm#5552]
https://psalm.dev/r/f3e22a8103

This should give an error for TypeDoesNotContainType, since it's always type string.
It seems to work sometimes (getting the error) but sometimes it doesn't.

Also the vice versa should work:
if Psalm is not sure that the type is always the same, it should require the third parameter to be true.
As currently it can happen that psalm gives an error for TypeDoesNotContainType, the dev removes "true" 3rd arg.
Then some changes happen, e.g. adding int types to the array - but the dev isn't aware of that and suddenly the condition is wrong. In these cases psalm should require true for the 3rd arg.

**Comment by mathroc** [comment]
Hi, I think this issue is related to something I'm seeing: 

https://psalm.dev/r/6a0cefcf0f

↑ here I would expect this error instead (or something similar) on the if condition:

> ERROR: TypeDoesNotContainType - 11:5 - Operand of type false is always false

---

would creating a return type provider for in_array fix this issue too?

**Comment by orklah** [comment]
InvalidArrayOffset is not incorrect here, and adding the return type provider could handle this and emit a TypeDoesNotContainType  also, but the first one will stay.

This is related to two different systems:
- The first one is identifying the return type of in_array(), the return type provider would help on that
- The second one is narrowing down the type of $key. In this case, we know that $key is one of the values of $a and $a is an empty array. So $key can only be `empty`, in the sense of `never`. This is Psalm's talk for *this value can never happen*

Unfortunately, we have a better issue for that called `NoValue` which would be more appropriate than InvalidArrayOffset (it's the general case for when you try to add an absurd type as an offset, like another array).

It should not really happen on legit code though, so your snippet can't be representative of actual working code

**Comment by weirdan** [comment]
> the return type provider would help on that

but a stub with conditional return could be easier to implement and support.

**Comment by mathroc** [comment]
> InvalidArrayOffset is not incorrect here, and adding the return type provider could handle this and emit a TypeDoesNotContainType also, but the first one will stay.

Ok, that would be fine too. Having the TypeDoesNotContainType error would make it much more easy to understand

> but a stub with conditional return could be easier to implement and support.

Do you mean something in https://github.com/vimeo/psalm/blob/master/stubs/CoreGenericFunctions.phpstub ?

**Comment by weirdan** [comment]
> Do you mean something in master/stubs/CoreGenericFunctions.phpstub ?

Yeah.

**Comment by mathroc** [comment]
I'm unsure about how to do it with a stub, I tried that:

```php
/**
 * @psalm-template TNeedle
 * @psalm-template THayStackValue
 *
 * @param TNeedle $needle
 * @param array<THayStackValue> $haystack
 *
 * @return (TNeedle & THayStackValue is empty ? false: bool): bool)
 * @psalm-pure
 */
function in_array(mixed $needle, array $haystack, bool $strict = false)
{
}
```

obviously `TNeedle & THayStackValue` does not work there. is there some syntax I don't know about that would get me the correct type ?

> It should not really happen on legit code though, so your snippet can't be representative of actual working code

It is something that I found in actual working code :) with the exception the array is not always empty but Psalm think it is (I'll open another issue for that if one does not exists already)

**Comment by orklah** [comment]
> obviously TNeedle & THayStackValue does not work there. is there some syntax I don't know about that would get me the correct type ?

conditional types can be nested, for example: 
`(T is string ? string : (T is bool ? bool : mixed))` 

> I'll open another issue for that

please do :)

**Comment by mathroc** [comment]
> conditional types can be nested, for example:

I know, but wouldn't I have to make an enumeration of all possible types? (and their unions / intersections)

This feels like an infinitely long task :sweat_smile:

**Comment by orklah** [comment]
I think Bruce suggested this because your example can be solve simply with
`($haystack is array<empty, empty> ? false : bool)`

But if your actual code involves other types for $haystack, it may not be that easy, we'd need an example. But anyway, to resolve the original issue, I think a type provider will be necessary

**Comment by mathroc** [comment]
Ah, ok! Yes, this would be enough for my example, but I was hoping to solve this more generally :)

I'll look into the type providers
